In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
import numpy as np
!pip install mord
from mord import LogisticAT
from sklearn.preprocessing import StandardScaler
#불러오기
abalone= pd.read_csv("https://www.gagolewski.com/resources/data/ordinal-regression/abalone_ord.csv")
boston=pd.read_csv("https://www.gagolewski.com/resources/data/ordinal-regression/bostonhousing_ord.csv")

In [ ]:
#살펴보기
print(abalone)
print(boston)

      response  V1  V2  V3     V4     V5     V6      V7      V8      V9     V10
0            1   0   0   1  0.075  0.055  0.010  0.0020  0.0010  0.0005  0.0015
1            1   0   0   1  0.130  0.100  0.030  0.0130  0.0045  0.0030  0.0040
2            1   0   0   1  0.110  0.090  0.030  0.0080  0.0025  0.0020  0.0030
3            1   0   0   1  0.165  0.120  0.030  0.0215  0.0070  0.0050  0.0050
4            1   1   0   0  0.210  0.150  0.050  0.0385  0.0155  0.0085  0.0100
...        ...  ..  ..  ..    ...    ...    ...     ...     ...     ...     ...
4172         9   0   1   0  0.700  0.540  0.215  1.9780  0.6675  0.3125  0.7100
4173         9   1   0   0  0.690  0.540  0.185  1.6195  0.5330  0.3530  0.5550
4174        10   0   1   0  0.700  0.585  0.185  1.8075  0.7055  0.3215  0.4750
4175        10   1   0   0  0.665  0.535  0.225  2.1835  0.7535  0.3910  0.8850
4176        10   0   1   0  0.550  0.465  0.180  1.2125  0.3245  0.2050  0.5250

[4177 rows x 11 columns]
     response 

In [ ]:
#x,y로 데이터 나누어주기
x1=abalone.drop('response',axis=1)
y1=abalone['response']
x2=boston.drop('response',axis=1)
y2=boston['response']

In [ ]:
#비교를 위해 전체 데이터셋을 7:3의 비율로 training dataset와 test dataset으로 나누기.
X1_train, X1_test, y1_train, y1_test= train_test_split(x1,y1,test_size=0.3,random_state=903, stratify=y1)
X2_train, X2_test, y2_train, y2_test= train_test_split(x2,y2,test_size=0.3,random_state=903, stratify=y2)

#abalone

In [ ]:
#multinominal logistic regression
model= LogisticRegression(multi_class='multinomial')
model.fit(X1_train,y1_train)
y_pred1 = model.predict(X1_test)
print('정확도 :', metrics.accuracy_score(y1_test, y_pred1))


정확도 : 0.539872408293461


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [ ]:
#ordinal logistic regression
regressor = LogisticAT(alpha=1.0, verbose=0)
regressor.fit(X1_train,y1_train)
ypred=regressor.predict(X1_test)
print('정확도 :', metrics.accuracy_score(y1_test, ypred))

정확도 : 0.5303030303030303


#boston

In [ ]:
#multinomial logistic regression
model2= LogisticRegression(multi_class='multinomial')
model.fit(X2_train,y2_train)
y_pred2 = model.predict(X2_test)
print('정확도 :', metrics.accuracy_score(y2_test, y_pred2))


정확도 : 0.6447368421052632


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [ ]:
#ordinal logistic regression
regressor2 = LogisticAT(alpha=1.0, verbose=0)
regressor2.fit(X2_train,y2_train)
ypred2=regressor2.predict(X2_test)
print('정확도 :', metrics.accuracy_score(y2_test, ypred2))

정확도 : 0.7960526315789473


abalone 데이터에서는 multinomial logistic regression을 실행했을 때 정확도가 0.539872408293461, ordinal logistic regression을 실행했을 때에는 0.5303030303030303로 multinomial logistic regression이 더 높은 정확도를 가지므로 예측성능이 개선되었습니다.

또한 boston 데이터에서는 multinomial logistic regression을 실행했을 때 정확도가 0.6447368421052632, ordinal logistic regression을 실행했을 때에는 0.7960526315789473로 ordinal logistic regression이 더 높은 정확도를 보였습니다. 따라서 ordinal logistic regression이 더 좋은 모델이라고 할 수 있습니다.